core-rust/state-manager/src/receipt.rs

```diff
@@ -119,7 +126,7 @@ pub struct LocalTransactionReceipt {
 #[derive(Debug, Clone, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
 pub struct LedgerTransactionReceipt {
     pub outcome: LedgerTransactionOutcome,
-    pub substate_changes: SubstateChanges,
+    pub substate_changes: Vec<SubstateChange>,
```

Comment:
I see what you mean abstraction-wise, and you are right.

To justify my reasoning a little bit more:
- We claim that the `LedgerTransactionReceipt` is deterministic, subject to be merklized, etc., so there is some weak point for having an order here;
- With a `Vec<>` it is very clear how the substate change root hash is computed (i.e. you take hash of each element).
  - while with a map, one could ask "do I take the `(key, value)` or just `value`?"

We could use the `IndexMap` (or even the `BTreeMap`, which actually would give us the same ordering we want). But AFAIU, this would only be needed to optimize access to particular substates in some future Node logic, right? I would say: when this happens, we will probably need to introduce some intermediate "logic only" struct. Because the current `LedgerTransactionReceipt` is more focused on "being stored".